paddle/framework/net_design.md

```diff
@@ -0,0 +1,249 @@
+  virtual bool InferShape(Scope *scope) = 0;
+
+  static std::unique_ptr<NetBase> Create(const NetDef &def = NetDef());
+};
+```
+
+All network implementations should build networks from a protobuf message which 
+describes the structure of a real network; `Run` method should be implemented by 
+all implementations to offer a universal method to forward or backward compute a network.
+
+`NetBase::Create` is a method of factory pattern and can be implemented like
+
+```c++
+std::unique<NetBase> NetBase::Create(const NetDef& def) {
+  switch (def.model_type()) {
+    case NN:
+      return new Network(def);
+    case Recursive:
+      return new RecursiveNet(def);
+    case Recurrent:
+      return new RecurrentNet(def);
+  }
+  return nullptr;
+}
+```
+
+Network is designed as the container of operators, to make it more extendable,
+we decoupling it from the related variable resources. 
+
+`Run(Scope* scope)` takes the scope as a argument so that it can run in different scopes.
+
+Finally, `NetBase` can be used as followed
+
+```c++
+Scope default_scope;
+auto net = NetBase::CreateNet(def);
+
+if (net) {
+  net.Run(&default_scope);
+}
+```
+
+## `PlainNet` as a simple implementation of `BaseNet`
+
+A very basic implementation is as followed, all it does is simply to run every operators in sequence.
+
+```c++
+class PlainNet final : public NetBase {
+ public:
+  // Create a network describe by `def`.  NetDef is the definition of a network.
+  PlainNet(const NetDef &def);
+
+  virtual bool InferShape(Scope *scope) override;
+
+  // Run all the operators with the `scope`, if no scope is provided, default
+  // scope will be used instead.
+  virtual bool Run(Scope *scope = nullptr, OpIndex begin = -1,
+                   OpIndex end = -1) const override;
+
+  virtual OpIndex AddOp(const proto::OpDef &def) override;
+
+  virtual void AddOptimizerOps(const OptAttrs &attrs) override;
+
+  virtual void AddBackwardOps() override;
+
+ protected:
+  // Create operators accordding to `def`.
+  bool CreateNet(const NetDef &def);
+
+  // Add a operator which is identified as `type` and has attributes described
+  // in `attrs`, the `inputs` are the keys of readonly input variables,
+  // `outputs` are keys of mutable output variables. An `OpIndex` will be
+  // returned which indicates the offset of the new operator in `ops_`.
+  OpIndex AddOp(const std::string &type, const std::vector<string> &inputs,
+                const std::vector<string> &outputs,
+                const OprAttr &attrs = OprAttr());
+
+ private:
+  // the operators owned by `Network`.
+  std::vector<Operator> ops_;
+};
+```
+
+`PlainNet` will create operators so that a private member `ops_` is defined,
+the operators are created by `CreateNet`, and each operator is created by `AddOp`.
+
+
+## PlainNet Usage
+`PlainNet` can be used to define and run a network as followed
+
+```c++
+// create an empty scope located on CPU device.
+Scope scope(CPUPlace());
+
```

Comment:
In the offline discussion, we decide to create a context inside the net run method. Then, the context will pass to every operator's run method.